services/wallet/transfer/block_ranges_sequential_dao.go

```diff
@@ -58,85 +51,84 @@ func (b *BlockRangeSequentialDAO) getBlockRange(chainID uint64, address common.A
-}
-
-//lint:ignore U1000 Ignore unused function temporarily, TODO use it when new transfers are fetched
-func (b *BlockRangeSequentialDAO) updateLastBlock(chainID uint64, account common.Address, block *big.Int) (err error) {
-	return updateBlock(b.db, chainID, account, lastBlockColumn, block)
-}
-
-func (b *BlockRangeSequentialDAO) updateFirstBlock(chainID uint64, account common.Address, block *big.Int) (err error) {
-	return updateBlock(b.db, chainID, account, firstBlockColumn, block)
-}
-
-func updateBlock(creator statementCreator, chainID uint64, account common.Address,
-	blockColumn string, block *big.Int) (err error) {
+func (b *BlockRangeSequentialDAO) upsertRange(chainID uint64, account common.Address,
+	newBlockRange *BlockRange) (err error) {
 
-	update, err := creator.Prepare(fmt.Sprintf(`UPDATE blocks_ranges_sequential
-                SET %s = ?
-                WHERE address = ?
-                AND network_id = ?`, blockColumn))
+	log.Debug("upsert blocks range", "account", account, "chainID", chainID,
+		"start", newBlockRange.Start, "first", newBlockRange.FirstKnown, "last", newBlockRange.LastKnown)
 
+	blockRange, err := b.getBlockRange(chainID, account)
 	if err != nil {
 		return err
 	}
 
-	_, err = update.Exec((*bigint.SQLBigInt)(block), account, chainID)
-
-	if err != nil {
-		return err
-	}
+	// Update existing range
+	if blockRange != nil {
+		// Ovewrite start block if there was not any or if new one is older, because it can be precised only
+		// to a greater value, because no history can be before some block that is considered
+		// as a start of history, but due to concurrent block range checks, a newer greater block
+		// can be found that matches criteria of a start block (nonce is zero, balances are equal)
+		if newBlockRange.Start != nil || (blockRange.Start != nil && newBlockRange.Start != nil &&
+			blockRange.Start.Cmp(newBlockRange.Start) < 0) {
+			blockRange.Start = newBlockRange.Start
+		}
 
-	return
-}
+		// Overwrite first known block if there was not any or if new one is older
+		if (blockRange.FirstKnown == nil && newBlockRange.FirstKnown != nil) ||
+			(blockRange.FirstKnown != nil && newBlockRange.FirstKnown != nil && blockRange.FirstKnown.Cmp(newBlockRange.FirstKnown) > 0) {
+			blockRange.FirstKnown = newBlockRange.FirstKnown
+		}
 
-func (b *BlockRangeSequentialDAO) upsertRange(chainID uint64, account common.Address,
-	start *big.Int, first *big.Int, last *big.Int) (err error) {
+		// Overwrite last known block if there was not any or if new one is newer
+		if (blockRange.LastKnown == nil && newBlockRange.LastKnown != nil) ||
+			(blockRange.LastKnown != nil && newBlockRange.LastKnown != nil && blockRange.LastKnown.Cmp(newBlockRange.LastKnown) < 0) {
+			blockRange.LastKnown = newBlockRange.LastKnown
+		}
 
-	log.Info("upsert blocks range", "account", account, "network id", chainID, "start", start, "first", first, "last", last)
+		update, err := b.db.Prepare(`UPDATE blocks_ranges_sequential
```

Comment:
Can't we simplify these two separate update and insert operations into one using `INSERT OR REPLACE INTO blocks_ranges_sequential ...` given that the composite primary key will take care of the WHERE clause? Don't remember if our sqlite supports it though.